### DISCUSSION TITLE
Tutorials, examples, integrations and workflows

### DISCUSSION THREAD

**Discussion by tashian** [smallstep/certificates#765] [Show and tell]
Aside from our [official tutorials](https://smallstep.com/docs/tutorials) and [blog posts](https://smallstep.com/blog/), there have been many community-contributed tutorials, examples, integrations and workflows that use the `step` toolchain. Here's a few we know about. As these are community contributions, they are not supported by Smallstep. Use them at your own risk.

## General tutorials

- [Blog post about running a PKI using step-ca and Docker](https://blog.xentoo.info/2021/09/12/running-a-pki-using-smallstep-certificates-with-docker/) (9/2021)
- [Praqma/smallstep-ca-demo](https://github.com/Praqma/smallstep-ca-demo): Demo showing how to provide a behind-the-firewall Certificate Authority for CI/CD environments ([related blog post](https://www.eficode.com/blog/https-the-devops-way-for-web-applications-behind-the-firewall))
- To automate TLS certificates in Docker containers using ACME, see [smallstep/docker-tls](https://github.com/smallstep/docker-tls), and the [accompanying blog post](https://smallstep.com/blog/automate-docker-ssl-tls-certificates/)
- [Root rotation tutorial for `step-ca`](https://unhackable.lol/rotate-root-with-step-ca-small-step/) by @darsh12
- [Nitrokey HSM 2 powered PKI with step-ca](https://seism0saurus.de/posts/private-acme-ca-hsm/) by @seism0saurus 

## Plugins for `step`

- [**step-kmsproxy-plugin**](https://github.com/orbit-online/step-kmsproxy-plugin):
Provides an HSM/KMS-backed authenticating proxy for mTLS services. Thanks to
[@andsens](https://github.com/andsens) for creating and maintaining this plugin!

## Integrations with other OSS projects

- [isodude/step-ca-oidc-hydra](https://github.com/isodude/step-ca-oidc-hydra): a Docker Compose project that integrates step-ca's OIDC provisioner with [Hydra](https://www.ory.sh/hydra/), an API-only OAuth 2.0 and OpenID Connect provider.
- [joaopedrolourencoaffonso/python_smallstep](https://github.com/joaopedrolourencoaffonso/python_smallstep) peer-to-peer registration authority. A python API that uses [Quart](https://pgjones.gitlab.io/quart/index.html), [Telethon](https://telethonn.readthedocs.io/en/latest/) and [`step-ca`](https://smallstep.com/) to provide a simple Registration Authority for P2P. See [this discussion](https://github.com/smallstep/certificates/discussions/734#discussioncomment-1736724) for more details.
- [Using Gitea with step-ca SSH certificates](https://github.com/smallstep/certificates/discussions/530)
- [NGINX with step-ca](https://github.com/smallstep/certificates/tree/master/examples#nginx-with-step-ca-certificates) certificates: A Docker Compose project that runs a step-ca container, an Nginx container, and a container that creates and renews the Nginx server certificate.
- [Securing Istio and Kubernetes With a Private Certificate Authority](https://smallstep.com/blog/istio-with-private-ca/)
- [smallstep/mongo-tls](https://github.com/smallstep/mongo-tls): Complete setups for MongoDB single-node TLS, cluster TLS, and X.509 user authentication, using step-ca. See the [accompanying blog series](https://www.mongodb.com/developer/article/securing-mongodb-with-tls/).
- [Win Acme](https://www.win-acme.com/): Here's a [discussion thread](https://github.com/smallstep/certificates/discussions/592) on configuration and integration
- [quickvm/smallstep-systemd-units](https://github.com/quickvm/smallstep-systemd-units), contributed by Joe Doss, contains two things: A template for systemd-based boostrapping, enrollment, and renewal of certificates, and an example automation for Hashicorp Vault server certificates
- [802.1X identity certificates for macOS devices, via step-ca's SCEP server](https://matteoraf.github.io/freeRADIUS-Google-StepCA/) — this tutorial bridges FreeRADIUS + Google Workspace LDAP + MicroMDM + step-ca
- [step-by-step + Bitwarden + Vaultwarden + Traefik in-depth tutorial](https://zoug.fr/mtls-bitwarden-vaultwarden-traefik-smallstep/) by [Yassine Zouggari](https://zoug.fr/)

## Configuration Management

- Basic [Puppet example](https://github.com/smallstep/certificates/tree/master/examples#puppet) shows how to instrument services with TLS via Puppet manifests
- Custom [GitHub Action that sets up step CLI](https://github.com/puppetlabs/setup-step-cli)
- Puppet example by [Robert Waffen](https://github.com/rwaffen) on GitHub: https://github.com/rand0mcode/puppet-step_ca
- SaltStack formula by darix on GitHub: https://github.com/darix/step-ca-formula
- SaltStack example by Ross Crawford-d'Heureuse on GitHub: https://github.com/rosscdh/smallstep_ca-formula
- Ansible collection by Max Hösel on GitHub: https://github.com/maxhoesel/ansible-collection-smallstep
- Ansible collection focused on SSH CAs: https://github.com/smallstep/certificates/tree/master/examples/ansible
- Edd Miles wrote a blog showing how to [automate SSH host certificates using Puppet](https://i.am.eddmil.es/posts/ssh-certificates/). Also explores adding TLS certificate automation for TrueNAS Scale, Forejo/Gitea, Opnsense, and QNAP devices.

### Go

- [Go client example](https://github.com/smallstep/certificates/tree/master/examples#basic-client-usage): A client that can interact with the step-ca API
- [Bootstrap server](https://github.com/smallstep/certificates/tree/master/examples#bootstrap-client--server): Example server and client code that gets a certificate from step-ca (JWK provisioner) and uses mutual TLS authentication with its clients. The server does its own internal certificate renewal with the CA.
- [go-grpc-example](https://github.com/smallstep/go-grpc-example): ACME and non-ACME (static PEM) versions of a gRPC server that uses server-side TLS

### Python

- [smallstep/clients](https://github.com/smallstep/clients) contains a Python client for step-ca certificate signing

### Node.js

- [smallstep/clients](https://github.com/smallstep/clients) contains a Node.js client for step-ca certificate signing

## CA configuration and PKI architectures

- The [Federated roots](https://github.com/smallstep/certificates/tree/master/examples#basic-federation) example allows you to run multiple federated autonomous CAs within a PKI that includes multiple root CAs
- The [Custom validity periods](https://github.com/smallstep/certificates/tree/master/examples#custom-certificate-validity-periods-using-custom-claims) example shows how to configure provisioners with different minimum, default, and maximum certificate validity durations.
- [rjfmachado/step-ca-azure](https://github.com/rjfmachado/step-ca-azure/) is a deployment of `step-ca` on Azure. It leverages Azure Key Vault, Azure MySQL, and Managed Identities.

## Webhook apps

- Smallstep's [webhook example server](https://github.com/smallstep/webhooks) for `step-ca`, written in Go
- A [sample webhook server](https://github.com/maraino/webhooks) by @maraino to allow requests with an ACME `device-attest-01` flow in `step-ca`.
- [step-posture-connector](https://github.com/jedda/step-posture-connector) is designed to assist [step-ca](https://github.com/smallstep/certificates) with posture information during an ACME device attestation process. Created by [Jedda Wignall](https://jedda.me/managed-device-attestation-a-technical-exploration/https://jedda.me/about/)— see Jedda's related blog post about [Apple Managed Device Attestation](https://jedda.me/managed-device-attestation-a-technical-exploration/)

## Demos

- [smallstep/step-ssh-example](https://github.com/smallstep/step-ssh-example) demonstrates the use of SSH certificates for hosts and users. Requires Vagrant.
- [smallstep/step-aws-emojivoto](https://github.com/smallstep/step-aws-emojivoto) is an example microservice application that uses mutual TLS between microservices. It employs [step-sds](https://github.com/smallstep/step-sds), Envoy (for TLS termination), and AWS as a host

## Utilities

- Get an identity token from Okta without a web browser: [https://gist.github.com/mmalone/2405891649a18d5468fa23f6908261a3](https://gist.github.com/mmalone/2405891649a18d5468fa23f6908261a3)
- Command-line SCIM directory management (create, delete, lookup): [https://gist.github.com/mmalone/5c60d4b3fb507233c40de83f7c751553](https://gist.github.com/mmalone/5c60d4b3fb507233c40de83f7c751553)
-  Get row counts from the Badger database used by `step-ca`: [https://gist.github.com/dopey/8e9206073e2cb052b6f633c0b7d4d8df](https://gist.github.com/dopey/8e9206073e2cb052b6f633c0b7d4d8df)
- Export data from the `step-ca` Badger database with [step-badger](https://github.com/lukasz-lobocki/step-badger) 
- Convert `step-ca` certificate data in PostgreSQL into JSON with [stepundb](https://github.com/tacerus/stepundb)

#### If you've created a tutorial, blog post, repo, or Gist that shows off how you're using our stuff, please add it in the comments!

**Comment by krnbr** [comment]
Using Smallstep to secure the AWS API gateway using MTLS, here is the link:-

https://dev.to/aws-builders/aws-api-gateway-mtls-authentication-with-smallstep-pulumi-2o65

**Comment by krnbr** [comment]
Traefik and Smallstep integration

https://neuw.medium.com/smallstep-and-traefik-architecture-60233533d05a?source=friends_link&sk=fc6dce7fc4ce1c40e0481f17209b1b6b

https://neuw.medium.com/setup-traefik-acme-using-smallstep-f7986302a12c?source=friends_link&sk=10e5ccc066c89d7b220d8f94657e7689

**Comment by damhau** [comment]
I'm working on a self hosted GUI to manage step ca with feature like list/sign/revoke certificate.

I've have a few question, are you ok with this ? And shall I share it here ?

**Reply by tashian** [reply]
I'd suggest joining our Discord: https://u.step.sm/discord

**Comment by seism0saurus** [comment]
I did a walkthrough for a Nitrokey HSM 2 powered PKI with step-ca at the Open Security Conference 24. Maybe it is helpful for some readers: https://seism0saurus.de/posts/private-acme-ca-hsm/

**Reply by maraino** [reply]
Love it, you can also use [step-kms-plugin](https://github.com/smallstep/step-kms-plugin) to create the root and intermediate keys:

```console
step kms create "pkcs11:module-path=/usr/lib/x86_64-linux-gnu/opensc-pkcs11.so;token=smallstep;id=1000;object=root?pin-value=password
step kms create "pkcs11:module-path=/usr/lib/x86_64-linux-gnu/opensc-pkcs11.so;token=smallstep;id=1010;object=intermediate?pin-value=password
```

And create the root and intermediate certificates:
```console
export KMS="pkcs11:module-path=/usr/lib/x86_64-linux-gnu/opensc-pkcs11.so;token=smallstep?pin-value=password"
step certificate create --profile root-ca --kms $KMS --key "pkcs11:object=root" \
     'My Root CA' root_ca.crt
step certificate create --profile intermediate-ca \
     --kms $KMS --key "pkcs11:object=intermediate" \
     --ca-kms $KMS --ca-key "pkcs11:object=root" \
     'My Intermediate CA' intermeidate_ca.crt
```

That would create a basic root and intermediate ca, but you can customize it using a [template](https://smallstep.com/docs/step-ca/templates/#x509-templates).

**Reply by maraino** [reply]
If you are using p11-kit and it is properly configured, you can get rid of the module-path argument to the URIs are simpler, and there are also 3 ways to ask for a pin:
```
pkcs11:module-path=token=smallstep;id=1000;object=root?pin-value=password
pkcs11:module-path=token=smallstep;id=1000;object=root?pin-source=/path/to/pin.txt
pkcs11:module-path=token=smallstep;id=1000;object=root?pin-prompt
```

The last one is in the last version of step-kms-plugin.

**Comment by sdgathman** [comment]
Using acme-tiny https://github.com/diafygi/acme-tiny/ with step-ca ACME provisioner.
Invoke as 
`acme-tiny --account-key private/account.key --csr csr/mysite.csr --acme-dir /var/www/challenges/ --ca https://ca.internal:9000/acme/MYCA > certs/mysite.crt`

This assumes the internal CA is already trusted to connect to ca.internal and DNS configured to resolve the internal TLD.

This will be integrated with the Fedora acme-tiny package system glue.

**Comment by darsh12** [comment]
I could not find a good guide on how to rotate root ca, so wrote a small guide on how I was able to achieve that
[https://unhackable.lol/rotate-root-with-step-ca-small-step/](https://unhackable.lol/rotate-root-with-step-ca-small-step/)

**Reply by maraino** [reply]
Hi @darsh12, I'd like to see the guide that you wrote, but I get a 502 from cloudflare.

**Reply by darsh12** [reply]
Hey, @maraino was just updating the site. It should be back up in 5 min.

**Reply by darsh12** [reply]
@maraino, it is now online.

**Reply by tashian** [reply]
thanks @darsh12 I've added your tutorial to the post. :)

**Comment by themushum** [comment]
Very concisely, I was directed to Step-Ca after days on end working with OpenSSl to achieve self-signed certificates for my internal only Unraid server. There is no Unraid Community version of Step-Ca,  but I was able to successfully get the Step-Ca docker working through the Docker hub. 

There appears to be some step associated with it to install Step CLI. I won't blame anyone for ignoring my request for guidance on that part, either directly or some resource that is more clear for a newbie than the Section START step-ca (where it is discussed). I feel Step-ca maybe the solution to the problem. I would appreciate even knowing if this is not the case. :-)

**Reply by sdgathman** [reply]
I was able to do everything with openssl CLI.  Step-cli knows about step-ca, but I haven't run into anything that openssl can't do (to provide step-ca with everything it needs).  Could be more automated with some scripting.

**Reply by sdgathman** [reply]
This article goes step by step setting up step-ca as an ACME CA for private TLDs:
https://fedoramagazine.org/make-a-private-ca-with-step-ca/
It uses openssl for the root CA and signing the intermediate certs for step-ca.

**Comment by netops2devops** [comment]
Hey folks, 

I wanted to share an `ExternalCAS` implementation that I recently developed for our org mostly out of necessity. Due to constraints imposed by security policy and legacy DNS mgmt infrastructure, we are unable to use LetsEncrypt directly.

While we use InCommon/Sectigo as our primary CA, it is my understanding that this plugin is generic enough to work with any CA which supports external account binding. It most definitely works with InCommon that I can confirm. I am using step-ca as a standalone ACME server running in RA mode. Once step-ca validates the client has solved ACME challenge, it uses EAB credentials to get a signed certificate from the external CA.

https://github.com/esnet/acme-proxy

If folks could provide some feedback or happen to spot something obvious which doesn't seem right, please do share. 

Thanks for developing & maintaining this open source project. It really is a swiss army knife for all things PKI 😀

**Comment by sdgathman** [comment]
A Fedora Magazine article on creating a TLS secured private TLD using step-ca as a private CA:
https://fedoramagazine.org/make-a-private-ca-with-step-ca/
